File: backend/app/schemas/analytics.py
"""
Pydantic schemas for the Analytics domain.

Used by:
  GET /analytics/summary
  GET /analytics/by-category
  GET /analytics/monthly-trend
  GET /analytics/top-merchants
  GET /analytics/by-payment-mode
  GET /dashboard   (combined response)
"""
from calendar import month_abbr
from datetime import date
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, computed_field


# ── Individual response schemas ───────────────────────────────────────────────


class SummaryResponse(BaseModel):
    """High-level KPI numbers for a date range."""

    total_spent: Decimal
    tx_count: int
    avg_amount: Decimal
    largest_tx: Decimal
    period_from: date
    period_to: date


class CategoryBreakdownItem(BaseModel):
    """Spending total for one category."""

    category_name: str
    category_id: UUID | None = None
    total: Decimal
    count: int
    percentage: float   # share of period total (0–100)


class MonthlyTrendItem(BaseModel):
    """Spending total for one calendar month."""

    year: int
    month: int
    label: str          # e.g. "Jun 2025"
    total: Decimal
    count: int


class TopMerchantItem(BaseModel):
    """One merchant ranked by total spend."""

    merchant: str
    total: Decimal
    count: int


class PaymentModeItem(BaseModel):
    """Spending split for one payment mode."""

    payment_mode: str
    total: Decimal
    count: int
    percentage: float   # share of period total (0–100)


# ── Combined dashboard response ───────────────────────────────────────────────


class DashboardResponse(BaseModel):
    """
    All dashboard data in one response.

    Returned by GET /dashboard so the frontend makes a single request
    instead of five separate calls.
    """

    period_from: date
    period_to: date
    summary: SummaryResponse
    by_category: list[CategoryBreakdownItem]
    monthly_trend: list[MonthlyTrendItem]
    by_payment_mode: list[PaymentModeItem]
    top_merchants: list[TopMerchantItem]

File: frontend/src/api/analytics.js
import client from './client'

export const getDashboard = (params = {}) =>
  client.get('/analytics/dashboard', { params }).then(r => r.data)

export const getSummary = (params = {}) =>
  client.get('/analytics/summary', { params }).then(r => r.data)

export const getByCategory = (params = {}) =>
  client.get('/analytics/by-category', { params }).then(r => r.data)

export const getMonthlyTrend = (params = {}) =>
  client.get('/analytics/monthly-trend', { params }).then(r => r.data)

export const getTopMerchants = (params = {}) =>
  client.get('/analytics/top-merchants', { params }).then(r => r.data)

export const getByPaymentMode = (params = {}) =>
  client.get('/analytics/by-payment-mode', { params }).then(r => r.data)


// Phase 4 endpoints
export const getHealthScore = () =>
  client.get('/analytics/health-score').then(r => r.data)

export const getInsights = () =>
  client.get('/analytics/insights').then(r => r.data)
